augur/utils.py

```diff
@@ -143,66 +144,272 @@ def default(self, obj):
-                    if feature_names is None or fname in feature_names:
-                        features[fname] = feat
-                elif "gene" in feat.qualifiers:
+        return _read_genbank(reference, feature_names)
+
+def _read_nuc_annotation_from_gff(record, reference):
+    """
+    Looks for the 'record' & 'source' GFF type as well as the ##sequence-region
+    pragma. Note that 'source' isn't really a GFF feature type, but is used
+    widely in the Nextstrain ecosystem. If there are multiple we check that the
+    coordinates agree.
+    
+    Parameters
+    ----------
+    record : <class 'Bio.SeqRecord.SeqRecord'> reference: string
+
+    Returns
+    -------
+    <class 'Bio.SeqFeature.SeqFeature'>
+
+    Raises
+    ------
+    AugurError
+        If no information on the genome / seqid length is available or if the
+        information is contradictory
+    """
+    nuc = {}
+    # Attempt to parse the sequence-region pragma to learn the genome
+    # length (in the absence of record/source we'll use this for 'nuc')
+    sequence_regions = record.annotations.get('sequence-region', [])
+    if len(sequence_regions)>1:
+        raise AugurError(f"Reference {reference!r} contains multiple ##sequence-region pragma lines. Augur can only handle GFF files with a single one.")
+    elif sequence_regions:
+        from Bio.SeqFeature import SeqFeature, FeatureLocation
+        (name, start, stop) = sequence_regions[0]
+        nuc['pragma'] = SeqFeature(
+            FeatureLocation(start, stop),
+            strand=1,
+            type='##sequence-region pragma',
+            id=name,
+        )
+    for feat in record.features:
+        if feat.type == "region":
+            nuc['region'] = feat
+        elif feat.type == "source":
+            nuc['source'] = feat
+
+    if not nuc:
+        raise AugurError(f"Reference {reference!r} didn't define any information we can use to create the 'nuc' annotation. You can use a line with a 'record' or 'source' GFF type or a ##sequence-region pragma.")
+
+    # ensure they all agree on coordinates, if there are multiple
+    if len(nuc.values())>1:
+        coords = [(name, int(feat.location.start), int(feat.location.end)) for name,feat in nuc.items()]
+        if not all(el[1]==coords[0][1] and el[2]==coords[0][2] for el in coords):
+            raise AugurError(f"Reference {reference!r} contained contradictory coordinates for the seqid/genome. We parsed the following coordinates: " + 
+                             ', '.join([f"{el[0]}: [{el[1]+1}, {el[2]}]" for el in coords]) # +1 on the first coord to shift to one-based GFF representation
+                             )
+
+    if 'region' in nuc:
+        return nuc['region']
+    if 'pragma' in nuc:
+        return nuc['pragma']
+    if 'source' in nuc:
+        return nuc['source']
+    raise AugurError("Internal error (_read_nuc_annotation_from_gff fallthrough). This is a bug in Augur's code.")
```

Comment:
It seems like we could replace this exception with the exception associated with an empty `nuc` dictionary here and use an `if`/`elif`/`else` structure to choose what to return like so:

```python
if 'region' in nuc:
    return nuc['region']
elif 'pragma' in nuc:
    return nuc['pragma']
elif 'source' in nuc:
    return nuc['source']
else:
    raise AugurError(f"Reference {reference!r} didn't define any information we can use to create the 'nuc' annotation. You can use a line with a 'record' or 'source' GFF type or a ##sequence-region pragma.")
```

We shouldn't need a "fallthrough" exception then.

Reply:
Fixed (part of this [diff](https://github.com/nextstrain/augur/compare/17f76c184fd86bf2bccb3a2f1d1ad62384d33d0a..95e427a1a683a1d24e29b62bee7b5be3a56b01a0)]